File: src/main/java/com/pedigree/gedcom/GedcomMapper.java
package com.pedigree.gedcom;

import com.pedigree.model.Family;
import com.pedigree.model.Gender;
import com.pedigree.model.Individual;

import java.time.LocalDate;
import java.time.Month;
import java.time.format.TextStyle;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

public class GedcomMapper {
    public static Map<String, String> buildIndividualXrefs(List<Individual> individuals) {
        Map<String, String> map = new HashMap<>();
        for (int i = 0; i < individuals.size(); i++) {
            map.put(individuals.get(i).getId(), "@I" + (i + 1) + "@");
        }
        return map;
    }

    public static Map<String, String> buildFamilyXrefs(List<Family> families) {
        Map<String, String> map = new HashMap<>();
        for (int i = 0; i < families.size(); i++) {
            map.put(families.get(i).getId(), "@F" + (i + 1) + "@");
        }
        return map;
    }

    public static String buildName(Individual ind) {
        String first = safe(ind.getFirstName());
        String last = safe(ind.getLastName());
        // GEDCOM NAME format: Given /Surname/
        return (first + " /" + last + "/").trim();
    }

    public static String sexCode(Gender gender) {
        if (gender == null) return "U";
        return switch (gender) {
            case MALE -> "M";
            case FEMALE -> "F";
            default -> "U";
        };
    }

    public static Gender parseSex(String code) {
        if (code == null) return Gender.UNKNOWN;
        return switch (code.trim().toUpperCase(Locale.ROOT)) {
            case "M" -> Gender.MALE;
            case "F" -> Gender.FEMALE;
            default -> Gender.UNKNOWN;
        };
    }

    public static String formatDate(LocalDate date) {
        if (date == null) return null;
        // Use day month-abbrev year, e.g., 5 JAN 1980 per GEDCOM 5.5
        String mon = date.getMonth().getDisplayName(TextStyle.SHORT, Locale.ENGLISH).toUpperCase(Locale.ROOT);
        // Ensure 3-letter month abbreviations
        if (mon.length() > 3) mon = mon.substring(0, 3);
        return date.getDayOfMonth() + " " + mon + " " + date.getYear();
    }

    public static LocalDate parseDate(String text) {
        if (text == null) return null;
        String s = text.trim().toUpperCase(Locale.ROOT);
        // Expect formats like: 5 JAN 1980, 01 JAN 1980, JAN 1980, 1980
        String[] parts = s.split("\\s+");
        try {
            if (parts.length == 3) {
                int day = Integer.parseInt(parts[0]);
                Month m = parseMonth(parts[1]);
                int year = Integer.parseInt(parts[2]);
                if (m != null) return LocalDate.of(year, m, Math.min(Math.max(1, day), 28));
            } else if (parts.length == 2) {
                Month m = parseMonth(parts[0]);
                int year = Integer.parseInt(parts[1]);
                if (m != null) return LocalDate.of(year, m, 1);
            } else if (parts.length == 1) {
                int year = Integer.parseInt(parts[0]);
                return LocalDate.of(year, 1, 1);
            }
        } catch (Exception ignore) {
            return null;
        }
        return null;
    }

    private static Month parseMonth(String mon) {
        if (mon == null) return null;
        String m = mon.trim().toUpperCase(Locale.ROOT);
        // Accept 3-letter English abbreviations
        return switch (m) {
            case "JAN" -> Month.JANUARY;
            case "FEB" -> Month.FEBRUARY;
            case "MAR" -> Month.MARCH;
            case "APR" -> Month.APRIL;
            case "MAY" -> Month.MAY;
            case "JUN" -> Month.JUNE;
            case "JUL" -> Month.JULY;
            case "AUG" -> Month.AUGUST;
            case "SEP", "SEPT" -> Month.SEPTEMBER;
            case "OCT" -> Month.OCTOBER;
            case "NOV" -> Month.NOVEMBER;
            case "DEC" -> Month.DECEMBER;
            default -> null;
        };
    }

    public static String safe(String s) {
        return s == null ? "" : s;
    }

    public static String[] parseName(String nameLine) {
        // Returns [given, surname]
        if (nameLine == null) return new String[]{"", ""};
        String s = nameLine.trim();
        int firstSlash = s.indexOf('/');
        int secondSlash = (firstSlash >= 0) ? s.indexOf('/', firstSlash + 1) : -1;
        String given = s;
        String surname = "";
        if (firstSlash >= 0 && secondSlash > firstSlash) {
            given = s.substring(0, firstSlash).trim();
            surname = s.substring(firstSlash + 1, secondSlash).trim();
        }
        return new String[]{given, surname};
    }
}

File: src/main/kotlin/com/family/tree/gedcom/GedcomMapper.kt
package com.family.tree.gedcom

import com.family.tree.model.Family
import com.family.tree.model.Gender
import com.family.tree.model.Individual
import java.time.LocalDate
import java.time.Month
import java.time.format.TextStyle
import java.util.Locale

/**
 * Kotlin port of GedcomMapper (1:1 with Java implementation).
 */
object GedcomMapper {
    @JvmStatic
    fun buildIndividualXrefs(individuals: List<Individual>): Map<String, String> {
        val map = HashMap<String, String>()
        for (i in individuals.indices) {
            map[individuals[i].id] = "@I${i + 1}@"
        }
        return map
    }

    @JvmStatic
    fun buildFamilyXrefs(families: List<Family>): Map<String, String> {
        val map = HashMap<String, String>()
        for (i in families.indices) {
            map[families[i].id] = "@F${i + 1}@"
        }
        return map
    }

    @JvmStatic
    fun buildName(ind: Individual): String {
        val first = safe(ind.firstName)
        val last = safe(ind.lastName)
        // GEDCOM NAME format: Given /Surname/
        return "$first /$last/".trim()
    }

    @JvmStatic
    fun sexCode(gender: Gender?): String {
        if (gender == null) return "U"
        return when (gender) {
            Gender.MALE -> "M"
            Gender.FEMALE -> "F"
            else -> "U"
        }
    }

    @JvmStatic
    fun parseSex(code: String?): Gender {
        if (code == null) return Gender.UNKNOWN
        return when (code.trim().uppercase(Locale.ROOT)) {
            "M" -> Gender.MALE
            "F" -> Gender.FEMALE
            else -> Gender.UNKNOWN
        }
    }

    @JvmStatic
    fun formatDate(date: LocalDate?): String? {
        if (date == null) return null
        // Use day month-abbrev year, e.g., 5 JAN 1980 per GEDCOM 5.5
        var mon = date.month.getDisplayName(TextStyle.SHORT, Locale.ENGLISH).uppercase(Locale.ROOT)
        // Ensure 3-letter month abbreviations
        if (mon.length > 3) mon = mon.substring(0, 3)
        return "${date.dayOfMonth} $mon ${date.year}"
    }

    @JvmStatic
    fun parseDate(text: String?): LocalDate? {
        if (text == null) return null
        val s = text.trim().uppercase(Locale.ROOT)
        // Expect formats like: 5 JAN 1980, 01 JAN 1980, JAN 1980, 1980
        val parts = s.split("\u0020", "\t", "\n", "\r").filter { it.isNotEmpty() }
        return try {
            when (parts.size) {
                3 -> {
                    val day = parts[0].toInt()
                    val m = parseMonth(parts[1])
                    val year = parts[2].toInt()
                    if (m != null) LocalDate.of(year, m, kotlin.math.min(kotlin.math.max(1, day), 28)) else null
                }
                2 -> {
                    val m = parseMonth(parts[0])
                    val year = parts[1].toInt()
                    if (m != null) LocalDate.of(year, m, 1) else null
                }
                1 -> {
                    val year = parts[0].toInt()
                    LocalDate.of(year, 1, 1)
                }
                else -> null
            }
        } catch (_: Exception) {
            null
        }
    }

    private fun parseMonth(mon: String?): Month? {
        if (mon == null) return null
        val m = mon.trim().uppercase(Locale.ROOT)
        return when (m) {
            "JAN" -> Month.JANUARY
            "FEB" -> Month.FEBRUARY
            "MAR" -> Month.MARCH
            "APR" -> Month.APRIL
            "MAY" -> Month.MAY
            "JUN" -> Month.JUNE
            "JUL" -> Month.JULY
            "AUG" -> Month.AUGUST
            "SEP", "SEPT" -> Month.SEPTEMBER
            "OCT" -> Month.OCTOBER
            "NOV" -> Month.NOVEMBER
            "DEC" -> Month.DECEMBER
            else -> null
        }
    }

    @JvmStatic
    fun safe(s: String?): String = s ?: ""

    @JvmStatic
    fun parseName(nameLine: String?): Array<String> {
        // Returns [given, surname]
        if (nameLine == null) return arrayOf("", "")
        val s = nameLine.trim()
        val firstSlash = s.indexOf('/')
        val secondSlash = if (firstSlash >= 0) s.indexOf('/', firstSlash + 1) else -1
        var given = s
        var surname = ""
        if (firstSlash >= 0 && secondSlash > firstSlash) {
            given = s.substring(0, firstSlash).trim()
            surname = s.substring(firstSlash + 1, secondSlash).trim()
        }
        return arrayOf(given, surname)
    }
}